Tweak shadow lieutenant warnings to remove excessive repeat popups

Inglonias opened the pull request:
<!-- HOW TO USE: Under each "#### Heading" below, enter information relevant to your pull request.
Leave the headings unless they don't apply to your PR.

Please read carefully and don't delete the comments delimited by "< !--" and "-- >"
Once a pull request is submitted, automatic stylistic and consistency checks will be performed on the PR's changes.
The results of these can be either seen under the "Files changed" section of a PR or in the check's details.

NOTE: Please grant permission for repository maintainers to edit your PR.  It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them. -->

#### Summary
Bugfixes "Fixes excessive shadow lieutenant warnings"
<!-- This section should consist of exactly one line, edit the one above.
1. Replace the word "Category" with one of these words: Features, Content, Interface, Mods, Balance, Bugfixes, Performance, Infrastructure, Build, I18N.
2. Replace the text inside the quotes with a brief description of your changes.
Or if you don't want a changelog entry, replace the whole line with just the word "None" (with no quotes).
For more on the meaning of each category, see:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CHANGELOG_GUIDELINES.md
If approved and merged, your summary will be added to the project changelog:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/changelog.txt -->

#### Purpose of change
Fixes #71672 

#### Describe the solution
You get a maximum of one warning before the 7 day mark (before when the monster can actually spawn), and after that, warnings will fire when it fails to spawn. The first time it doesn't spawn, you will get a popup warning. After that, warnings just go into the logs.

#### Describe alternatives you've considered
None

#### Testing
Created a character and had them stand around a field for a week and change. The shadow lieutenant EOCs fired as I expected

#### Additional context
Pinging @RenechCDDA as this is their baby. This okay?


<!--README: Cataclysm: Dark Days Ahead is released under the Creative Commons Attribution ShareAlike 3.0 license.
The code and content of the game is free to use, modify, and redistribute for any purpose whatsoever.
By contributing to the project you agree to the term of the license and that any contribution you make will also be covered by the same license.
See http://creativecommons.org/licenses/by-sa/3.0/ for details. -->

Inglonias committed:
Tweak shadow lieutenant warnings

You get one early warning, and one late warning. Tops.

Inglonias committed:
Fixed conditions

I _SAID_: One warning BEFORE 7 days. One warning AFTER 7 days.

RenechCDDA commented:
> You get one warning before 7 days, and possibly a second one after 7 days. You can see each group of snippets once.

Just confirming this works as described. 

Not intended as approval, I'd personally like to see some more discussion on the linked issue before anything is committed. But the PR works fine for its stated purpose.

Inglonias committed:
Remove two-warning maximum

One warning allowed before spawn time, as many late warnings as you want, though.

RenechCDDA commented:
OK that last commit was basically what I was going to suggest. So this limits warnings to one total before spawning, and keeps them uncapped otherwise (with warnings being a signal for "you just missed getting attacked by 5/6 chance"). Looks good to me

Inglonias committed:
Added a non-popup version of the warning

Non popup fires if you've seen the warning more than twice, total.

andrei8l commented:
Can you merge these EOCs instead of duplicating the conditions? So you'd have `EOC_LIEUTENANT_SHADOW_WARN` run the common `conditions`, then in the true effect run the discriminator condition ( `{ "math": [ "Shadow_Warnings", "<", "2" ] },` looks like?) that selects between popup or message.

Inglonias commented:
I would need two copies of that condition, which makes this ugly, but probably less ugly than how I have it.

andrei8l commented:
> I would need two copies of that condition, which makes this ugly, but probably less ugly than how I have it.

Not sure I understand. You think duplicating the entire condition block + boilerplate is better than duplicating a single condition?

<details>
<summary>Maybe I didn't express it clearly, this is what I meant</summary>

Note - I didn't test it

```diff
diff --git a/data/json/monsters/zed_lieutenant.json b/data/json/monsters/zed_lieutenant.json
index 2e097f480d..5e55bd5dee 100644
--- a/data/json/monsters/zed_lieutenant.json
+++ b/data/json/monsters/zed_lieutenant.json
@@ -364,7 +364,6 @@
         {
           "or": [ { "math": [ "Shadow_Lurking", ">", "0" ] }, { "not": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] } } ]
         },
-        { "math": [ "Shadow_Warnings", "<", "2" ] },
         {
           "or": [ { "math": [ "Shadow_Warnings", "<", "1" ] }, { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] } ]
         },
@@ -377,41 +376,17 @@
     "effect": [
       { "math": [ "Shadow_Warnings", "+=", "1" ] },
       {
-        "if": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] },
-        "then": { "u_message": "Shadow_Warnings_Snippets_late", "snippet": true, "popup": true },
-        "else": { "u_message": "Shadow_Warnings_Snippets_early", "snippet": true, "popup": true }
-      }
-    ],
-    "false_effect": [ { "run_eocs": [ "EOC_LIEUTENANT_SHADOW_WARN_LITE" ] } ]
-  },
-  {
-    "type": "effect_on_condition",
-    "id": "EOC_LIEUTENANT_SHADOW_WARN_LITE",
-    "eoc_type": "ACTIVATION",
-    "//1": "Same conditions as EOC_LIEUTENANT_SPAWN_SHADOW, minus the RNG check and the warnings check.",
-    "//2": "This doesn't put a popup message in front of you.",
-    "condition": {
-      "and": [
-        { "not": "is_day" },
-        { "not": { "u_has_effect": "sleep" } },
-        {
-          "or": [ { "math": [ "Shadow_Lurking", ">", "0" ] }, { "not": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] } } ]
-        },
-        {
-          "or": [ { "math": [ "Shadow_Warnings", "<", "1" ] }, { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] } ]
+        "if": { "math": [ "Shadow_Warnings < 2" ] },
+        "then": {
+          "if": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] },
+          "then": { "u_message": "Shadow_Warnings_Snippets_late", "snippet": true, "popup": true },
+          "else": { "u_message": "Shadow_Warnings_Snippets_early", "snippet": true, "popup": true }
         },
-        "u_is_outside",
-        {
-          "or": [ { "u_at_om_location": "field" }, { "u_at_om_location": "forest" }, { "u_at_om_location": "forest_thick" } ]
+        "else": {
+          "if": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] },
+          "then": { "u_message": "Shadow_Warnings_Snippets_late", "snippet": true, "popup": false },
+          "else": { "u_message": "Shadow_Warnings_Snippets_early", "snippet": true, "popup": false }
         }
-      ]
-    },
-    "effect": [
-      { "math": [ "Shadow_Warnings", "+=", "1" ] },
-      {
-        "if": { "math": [ "time_since('cataclysm', 'unit':'days') >= 7" ] },
-        "then": { "u_message": "Shadow_Warnings_Snippets_late", "snippet": true, "popup": false },
-        "else": { "u_message": "Shadow_Warnings_Snippets_early", "snippet": true, "popup": false }
       }
     ]
   },
```


</details>

Inglonias commented:
No, I understood you. I was just grumbling about it a little bit.

Inglonias committed:
Removed EOC_LIEUTENANT_SHADOW_WARN_LITE

Replaced it with a fancier EOC_LIEUTENANT_SHADOW_WARN

Inglonias commented:
I don't know if the patch that @andrei8l presented would work as presented. I had my own patch after a few minutes, and I tested it. Seems to work.
